substrate/client/network/src/request_responses.rs

```diff
@@ -594,6 +618,59 @@ impl NetworkBehaviour for RequestResponsesBehaviour {
 
 	fn poll(&mut self, cx: &mut Context) -> Poll<ToSwarm<Self::ToSwarm, THandlerInEvent<Self>>> {
 		'poll_all: loop {
+			// Poll the periodic request check.
+			if let Poll::Ready(_) = self.periodic_request_check.poll_tick(cx) {
+				// First pass is needed to collect timed out requests and
+				// submit `OutboundFailure::Timeout` on the moved oneshot sender.
+				let timedout_requests: Vec<_> = self
+					.pending_requests
+					.iter()
+					.filter_map(|(id, req)| {
+						let Some(ProtocolDetails { request_timeout, .. }) =
+							self.protocols.get(&id.protocol)
+						else {
+							log::warn!(
+								target: "sub-libp2p",
+								"Request {id:?} has no protocol registered.",
+							);
+							return Some(id.clone())
+						};
+
+						let elapsed = req.started_at.elapsed();
+						if elapsed > *request_timeout {
+							Some(id.clone())
+						} else {
+							None
+						}
+					})
+					.collect();
+
+				if !timedout_requests.is_empty() {
+					log::warn!(
+						target: "sub-libp2p",
+						"Requests {timedout_requests:?} force detected as timeout.",
+					);
+				}
+
+				for id in timedout_requests {
+					let Some(req) = self.pending_requests.remove(&id) else {
```

Comment:
This 

https://github.com/paritytech/polkadot-sdk/pull/7222/files#diff-052aeaf79fef3d9a18c2cfd67006aa306b8d52e848509d9077a6a0f2eb856af7L769

and this 

https://github.com/paritytech/polkadot-sdk/pull/7222/files#diff-052aeaf79fef3d9a18c2cfd67006aa306b8d52e848509d9077a6a0f2eb856af7L842

will trigger in case where the response arrive after we timed it out or if we race with the timeout from sub-libp2p, so we should fix them.

Reply:
Yep good point, have removed the debug asserts, as holding the request ID might grow over time if libp2p fails to give us back the responses